FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,46 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				if let cache = try ManagedFeedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				let managedFeed = ManagedFeedImage.feedImages(from: feed, in: context)
+				try ManagedFeedCache.insertNewUniqueInstance(with: timestamp, feed: managedFeed, in: context)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				try ManagedFeedCache.find(in: context).map(context.delete)
+				try context.save()
```

Comment:
In case there is no previously stored cache `context.delete` will not be called, thus  there will be no changes in the context and will be no need to save it. Otherwise, Core Data may perform unnecessary work.

Please move `context.save` call inside `if let ...` block

```suggestion
				if let cache = try ManagedFeedCache.find(in: context) {
					context.delete(cache)
					try context.save()
				}
```

Reply:
hi with this code, when context.save() is wrapped in "if let..." block  this test is failing "test_delete_deliversErrorOnDeletionError" :)

Reply:
Hi @xvladax!

We've updated the failing test, please pull the latest changes to your branch